### ISSUE TITLE
CLIP Guided Diffusion

### ISSUE THREAD

**Issue by alishbaimran** [alishbaimran/protein-bi-encoder#2]
Notebook notes: https://github.com/openai/glide-text2im/blob/main/notebooks/clip_guided.ipynb
- Base diffusion model generates low-resolution images (64x64 pixels) from text prompts. The base model starts with an initial noise sample drawn from a Gaussian distribution. Uses clip func guidance.
- Upsampler model refines low-resolution images to higher resolution (256x256 pixels). The upscaler model starts with the low-resolution image generated by the base model. Doens't use clip func guidance. 
- The guidance func adjusts the generation process based on feedback from the CLIP model

CLIP diffusion process (training):
- use a dataset of clean images and corresponding text descriptions.
- select a clean image and corresponding text description
- add Gaussian noise to the image 
- predict Noise: use the diffusion model to predict the noise component at step 𝑡
- encode noisy image and text prompt using CLIP model
- compute similarity between the encoded image and text features (dot product of embeddings) 
- compute grad and adjust predicted noise 
- compute the loss between the adjusted predicted noise and the actual noise
-update the model’s parameters using backpropagation

CLIP diffusion process (inference):
- start with an initial noise sample drawn from a Gaussian distribution
- the diffusion model predicts the noise component that should be removed at each step
- the image (still with noise) and text at that step it is encoded using the CLIP model.
- similarity between the encoded image and text features is computed. 
- gradient is calculated of the similarity score with respect to the noisy image.
- use the gradient to adjust the predicted noise (set guidance_scale strength of guidance from CLIP model)
- update current image state by removing the noise

The only difference in training that you start with actual images and corrupt them vs just noise in inference time, and for training you have loss compute and backprop. Initial attempt at combining here: https://github.com/alishbaimran/protein-bi-encoder/blob/main/diffusion.py
